jib-core/src/main/java/com/google/cloud/tools/jib/image/ImageLayers.java

```diff
@@ -19,34 +19,32 @@
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 import java.util.Iterator;
+import java.util.LinkedHashSet;
 import javax.annotation.Nullable;
 
-/** Holds the layers for an image. Makes sure that each layer is only added once. */
+/** Holds the layers for an image. Makes sure that there are no duplicate layers. */
 public class ImageLayers<T extends Layer> implements Iterable<T> {
 
   public static class Builder<T extends Layer> {
 
-    private final ImmutableList.Builder<T> layersBuilder = ImmutableList.builder();
+    private final LinkedHashSet<T> layers = new LinkedHashSet<>();
     private final ImmutableSet.Builder<DescriptorDigest> layerDigestsBuilder =
         ImmutableSet.builder();
 
-    /** The last layer added. */
-    @Nullable private T lastLayer;
-
     /**
-     * Adds a layer.
+     * Adds a layer. Removes any prior occurrences of the same layer.
+     *
+     * <p>Note that only subclasses of {@link Layer} that implement {@code equals/hashCode} will be
+     * guaranteed to not be duplicated.
      *
      * @param layer the layer to add
      * @return this
      * @throws LayerPropertyNotFoundException if adding the layer fails
      */
     public Builder<T> add(T layer) throws LayerPropertyNotFoundException {
-      // Doesn't add the layer if the last layer is the same.
-      if (!isSameAsLastLayer(layer)) {
-        layerDigestsBuilder.add(layer.getBlobDescriptor().getDigest());
-        layersBuilder.add(layer);
-        lastLayer = layer;
-      }
+      layerDigestsBuilder.add(layer.getBlobDescriptor().getDigest());
+      layers.remove(layer);
```

Comment:
Don't `LinkedHashSet`s already handle duplicates?

Reply:
Turns out this is to re-order the element (https://github.com/GoogleContainerTools/jib/pull/739#discussion_r206358351).

That said, it seems worth to put a comment like "remove necessary to rearrange the insertion-order of `LinkedHashSet`".